### DISCUSSION TITLE
How to use query builder with stored function returning `setof`?

### DISCUSSION THREAD

**Discussion by KazimirPodolski** [mikro-orm/mikro-orm#5982] [Q&A] [answered]
I have entities `Energy` and `System`, many `Energy` to one `System`. And a function returning `setof` of entity's table `energy`:
```sql
CREATE OR REPLACE FUNCTION energy_upscaled() RETURNS setof energy
AS
$$
...
```

I'm trying to write a query with query builder selecting from `energy_upscaled` and joining `system`, but getting errors:

1. This way 
    ```typescript
    em
      .createQueryBuilder(Energy, 'upscaled')
      .from<Energy>(em.raw('energy_upscaled()'))
      .join('system', 'system', { 'system.uuid': 'upscaled.system' })
    ```
    fails with `Metadata for entity undefined not found`
2. This way
    ```typescript
    em
      .createQueryBuilder(Energy, 'upscaled')
      .from<Energy>(em.raw('energy_upscaled()'), 'upscaled')
      .join('system', 'system', { 'system.uuid': 'upscaled.system' })
    ```
    fails with `Cannot override the alias to 'upscaled' since a query already contains references to 'upscaled'`
3. This way
    ```typescript
    em
      .createQueryBuilder(Energy)
      .from<Energy>(em.raw('energy_upscaled()'), 'upscaled')
      .join('system', 'system', { 'system.uuid': 'upscaled.system' })
    ```
    fails with `Cannot override the alias to 'upscaled' since a query already contains references to 'e0'`
4. This way
    ```typescript
    em
      .createQueryBuilder(Energy)
      .from<Energy>(em.raw('energy_upscaled()'))
      .join('system', 'system', { 'system.uuid': 'e0.system' })
    ```
    fails with `Metadata for entity undefined not found`

So, supposedly I have to have an alias for `FROM energy_upscaled()`, but I cannot find a way to write it with query builder. Am I missing something? Or is it not supported at all?

**Comment by B4nan** [comment, marked as answer]
You could try to create a virtual entity represented by `energy_upscaled()` query or something similar. Or just use knex if you want this kind of flexibility on the raw side of things.

> Or is it not supported at all?

The whole `qb.from()` is a community addition, the QB is designed to work with entities, so yes, I would call this unsupported. But maybe its just about a few lines of code we could adjust to support it. Maybe just try to comment out that validation for `fails with Cannot override the alias to 'upscaled' since a query already contains references to 'upscaled'` to see what happens, we could allow disabling that in some cases.

> .join('system', 'system', { 'system.uuid': 'upscaled.system' })

This also won't work, the condition parameter expects values, `upscaled.system` will be considered a string value like this. The ORM builds the join condition for you, this parameter is only for additions to the main PK-FK condition which is inferred from the relationship.

**Reply by KazimirPodolski** [reply]
> You could try to create a virtual entity represented by energy_upscaled() query or something similar. Or just use knex if you want this kind of flexibility on the raw side of things.

I ended up with knex because my query is dynamical, I cannot put it into entity definition. It works now, thanks.
